diff -urN openssh-7.1p1/configure.ac openssh-7.1p1.new/configure.ac
--- openssh-7.1p1/configure.ac	2015-08-21 13:49:03.000000000 +0900
+++ openssh-7.1p1.new/configure.ac	2015-10-01 09:44:06.131215233 +0900
@@ -202,9 +202,9 @@
 	OSSH_CHECK_CFLAG_COMPILE([-fno-strict-aliasing])
 	OSSH_CHECK_CFLAG_COMPILE([-D_FORTIFY_SOURCE=2])
     if test "x$use_toolchain_hardening" = "x1"; then
-	OSSH_CHECK_LDFLAG_LINK([-Wl,-z,relro])
-	OSSH_CHECK_LDFLAG_LINK([-Wl,-z,now])
-	OSSH_CHECK_LDFLAG_LINK([-Wl,-z,noexecstack])
+	# OSSH_CHECK_LDFLAG_LINK([-Wl,-z,relro])
+	# OSSH_CHECK_LDFLAG_LINK([-Wl,-z,now])
+	# OSSH_CHECK_LDFLAG_LINK([-Wl,-z,noexecstack])
 	# NB. -ftrapv expects certain support functions to be present in
 	# the compiler library (libgcc or similar) to detect integer operations
 	# that can overflow. We must check that the result of enabling it
