diff -urN strongswan-5.3.3.org/src/include/linux/netlink.h strongswan-5.3.3/src/include/linux/netlink.h
--- strongswan-5.3.3.org/src/include/linux/netlink.h	2015-09-30 18:48:25.636948202 +0900
+++ strongswan-5.3.3/src/include/linux/netlink.h	2015-09-30 18:47:47.232948580 +0900
@@ -32,6 +32,8 @@
 
 #define MAX_LINKS 32
 
+typedef	unsigned short __kernel_sa_family_t;
+
 struct sockaddr_nl {
 	__kernel_sa_family_t	nl_family;	/* AF_NETLINK	*/
 	unsigned short	nl_pad;		/* zero		*/
